HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,128 @@
+import pyspark
+import time
+
+sc = pyspark.SparkContext()
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt')
+page_names = sc.textFile("s3://Harvard-CS205/wikipedia/titles-sorted.txt") 
+name_index = page_names.zipWithIndex().map(lambda (x,y): (x, y+1 ))
+key_index = name_index.map(lambda (x,y): (y,x))
+#print name_index.take(5)
+
+########################################################################################
+#Construct the graph 
+def link_to_KV(s):
+    parent, children = s.split(': ')
+    children = [int(to) for to in children.split(' ')]
+    return (int(parent),( 999999, int(parent), children))
+
+g = links.map(lambda x : link_to_KV(x))
+
+########################################################################################
+#Define BFS function
+
+def bfs(g, v, end_v , max_iter):
+#g is an RDD represent the graph
+#each item in g is in the form (char name, [neighbor 1, neighbor 2, ...])
+	#num_of_node = sc.accumulator(0)
+#v is the starting point of this BFS
+	num_of_iter = 0
+	num_node_not_touch_start = 0
+	g = g.map(lambda x: (init_start_node(x, v ) ) )
+	flag = True
+	for i in range(max_iter):
+		#num_node_already_touch = num_of_node
+		num_of_iter = num_of_iter + 1
+		g = g.flatMap(lambda x: update(x, num_of_iter - 1, end_v))
+		g = g.reduceByKey(lambda x,y : reconstruct(x,y))
+		#num_new_touch = g.filter(lambda x: x[1][0] == num_of_iter).count()
```

Comment:
It looks like you didn't use .cache() inside your loops - did this ever cause you trouble? I'm curious, because I found that I needed to cache RDDs in order for them to retain their values across iterations.